pkg/api/api.go

```diff
@@ -285,48 +285,53 @@ func (a *API) RegisterQuerier(queryable storage.Queryable, engine *promql.Engine
 	legacyPromHandler := fakeRemoteAddr(cacheGenHeaderMiddleware.Wrap(legacyPromRouter))
 
-	a.registerRouteWithRouter(router, a.cfg.LegacyHTTPPrefix+"/api/v1/read", querier.RemoteReadHandler(queryable), true, "POST")
-	a.registerRouteWithRouter(router, a.cfg.LegacyHTTPPrefix+"/api/v1/query", legacyPromHandler, true, "GET", "POST")
-	a.registerRouteWithRouter(router, a.cfg.LegacyHTTPPrefix+"/api/v1/query_range", legacyPromHandler, true, "GET", "POST")
-	a.registerRouteWithRouter(router, a.cfg.LegacyHTTPPrefix+"/api/v1/labels", legacyPromHandler, true, "GET", "POST")
-	a.registerRouteWithRouter(router, a.cfg.LegacyHTTPPrefix+"/api/v1/label/{name}/values", legacyPromHandler, true, "GET")
-	a.registerRouteWithRouter(router, a.cfg.LegacyHTTPPrefix+"/api/v1/series", legacyPromHandler, true, "GET", "POST", "DELETE")
+	a.registerRouteWithRouter(router, a.cfg.LegacyHTTPPrefix+"/api/v1/read", querier.RemoteReadHandler(queryable), needsAuth, "POST")
+	a.registerRouteWithRouter(router, a.cfg.LegacyHTTPPrefix+"/api/v1/query", legacyPromHandler, needsAuth, "GET", "POST")
+	a.registerRouteWithRouter(router, a.cfg.LegacyHTTPPrefix+"/api/v1/query_range", legacyPromHandler, needsAuth, "GET", "POST")
+	a.registerRouteWithRouter(router, a.cfg.LegacyHTTPPrefix+"/api/v1/labels", legacyPromHandler, needsAuth, "GET", "POST")
+	a.registerRouteWithRouter(router, a.cfg.LegacyHTTPPrefix+"/api/v1/label/{name}/values", legacyPromHandler, needsAuth, "GET")
+	a.registerRouteWithRouter(router, a.cfg.LegacyHTTPPrefix+"/api/v1/series", legacyPromHandler, needsAuth, "GET", "POST", "DELETE")
 	//TODO(gotjosh): This custom handler is temporary until we're able to vendor the changes in:
 	// https://github.com/prometheus/prometheus/pull/7125/files
-	a.registerRouteWithRouter(router, a.cfg.LegacyHTTPPrefix+"/api/v1/metadata", querier.MetadataHandler(distributor), true, "GET")
+	a.registerRouteWithRouter(router, a.cfg.LegacyHTTPPrefix+"/api/v1/metadata", querier.MetadataHandler(distributor), needsAuth, "GET")
 
 	// if we have externally registered routes then we need to return the server handler
 	// so that we continue to use all standard middleware
 	if registerRoutesExternally {
 		return a.server.HTTPServer.Handler
 	}
 
-	// Since we have a new router and the request will not go trough the default server
+	// Since we have a new router and the request will not go through the default server
 	// HTTP middleware stack, we need to add a middleware to extract the trace context
 	// from the HTTP headers and inject it into the Go context.
-	return nethttp.MiddlewareFunc(opentracing.GlobalTracer(), router.ServeHTTP, nethttp.OperationNameFunc(func(r *http.Request) string {
+	// Additionally add use the weaveworks AuthenticateUser middleware to copy the OrgID header
+	// into the request context.
+	return middleware.AuthenticateUser.Wrap(nethttp.MiddlewareFunc(opentracing.GlobalTracer(), router.ServeHTTP, nethttp.OperationNameFunc(func(r *http.Request) string {
```

Comment:
I don't quite understand this whole change. We have introduced `needsAuth`, set it to `false` and then here we do wrap `middleware.AuthenticateUser` anyway (which is the default auth middleware, used in any case except when auth is disabled).

Reply:
Basically we want to override the auth using `cfg.HTTPAuthMiddleware`, but not have it apply to the internal querier.  However, we still need the "normal" auth to still apply to translate `X-Scope-OrgID` => org id in the go context.  The querier (and almost every Cortex component) relies on this.

This change is confusing.  Marking as draft while we discuss internally.